spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java

```diff
@@ -615,10 +609,14 @@ private Long getStartSnapshotId(Long startTimestamp) {
     }
   }
 
-  private Long getEndSnapshotId(Long endTimestamp) {
+  private Long getEndSnapshotId(Long startTimestamp, Long endTimestamp) {
     Long endSnapshotId = null;
     for (Snapshot snapshot : SnapshotUtil.currentAncestors(table)) {
-      if (snapshot.timestampMillis() <= endTimestamp) {
+      long ts = snapshot.timestampMillis();
+      if (startTimestamp != null && ts <= startTimestamp) {
+        break;
+      }
```

Comment:
I found a bit more confusing by moving the logic here. We are checking if start timestamp is larger than the current snapshot's one. Checking within an ancestor loop yields the same result, but seems unnecessary. Plus, it has to introduces the line 571, which is another burden for reader to understand.
```
          || (endSnapshotId == null && endTimestamp != null)
```